pkg/kv/kvserver/rangefeed_restarter.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+var rangefeedRestartInterval = settings.RegisterDurationSetting(
+	settings.SystemOnly,
+	"kv.rangefeed.restart_interval",
+	"minimum interval between rangefeed restarts to conform to new settings",
+	50*time.Millisecond,
+	settings.WithVisibility(settings.Reserved),
+)
+
+var rangefeedRestartAvailablePercent = settings.RegisterIntSetting(
+	settings.SystemOnly,
+	"kv.rangefeed.restart_quota_headroom",
+	"minimum percentage of rangefeed catchup quota that still allow processor "+
+		"reconfiguration restarts to proceed. set to >100 to disable eager restarts on "+
+		"processor type change",
+	50,
+	settings.WithVisibility(settings.Reserved),
+	settings.WithValidateInt(func(i int64) error {
+		if i < 0 {
+			return errors.New("kv.rangefeed.restart_headroom must be equal or greater than zero")
+		}
+		return nil
+	}),
+)
+
+type RangefeedRestarter struct {
+	// Number of active catch up scans running.
+	// Note this value is updated synchronously with corresponding store limiter.
+	activeCatchups int64
+	s              *Store
+	st             *cluster.Settings
+	cfgChangedC    chan interface{}
+
+	// Cancel previous restart job context.
+	cancelJob func()
+}
+
+func NewRangefeedRestarter(s *Store, st *cluster.Settings) *RangefeedRestarter {
+	return &RangefeedRestarter{
+		s:           s,
+		st:          st,
+		cfgChangedC: make(chan interface{}, 1),
+		cancelJob:   func() {},
+	}
+}
+
+func (p *RangefeedRestarter) Start(ctx context.Context, stopper *stop.Stopper) {
+	RangeFeedUseScheduler.SetOnChange(&p.st.SV, func(context.Context) {
+		useScheduler := RangeFeedUseScheduler.Get(&p.st.SV)
+		p.start(ctx, useScheduler, stopper)
+	})
+	pacingChanged := func(ctx2 context.Context) {
+		select {
+		case p.cfgChangedC <- struct{}{}:
+		default:
+		}
+	}
+	concurrentRangefeedItersLimit.SetOnChange(&p.st.SV, pacingChanged)
+	rangefeedRestartInterval.SetOnChange(&p.st.SV, pacingChanged)
+	rangefeedRestartAvailablePercent.SetOnChange(&p.st.SV, pacingChanged)
+}
+
+// Cancel previous job and start new.
+func (p *RangefeedRestarter) start(ctx context.Context, toScheduler bool, stopper *stop.Stopper) {
+	p.cancelJob()
+	ctx, p.cancelJob = stopper.WithCancelOnQuiesce(ctx)
+	_ = stopper.RunAsyncTask(ctx, "restart range feed processors", func(ctx context.Context) {
+		run(ctx, p.s, toScheduler,
+			func(ctx context.Context) error {
+				return waitRestartQuota(ctx, timeutil.DefaultTimeSource{}, &p.activeCatchups, p.st, p.cfgChangedC)
+			},
+			func(r *Replica, p rangefeed.Processor) {
+				r.disconnectRangefeedWithErr(p,
+					kvpb.NewError(kvpb.NewRangeFeedRetryError(kvpb.RangeFeedRetryError_REASON_REPLICA_REMOVED)))
```

Comment:
`RangeFeedRetryError_REASON_RANGEFEED_CLOSED` is such an ambiguous result. It is not supposed to be retryable because we return this when client asked us to close the feed, but it is still a retry reason.

Reply:
What do you mean? `REASON_RANGEFEED_CLOSED` basically just means that the server closed the rangefeed. It's currently only used when the client requested it to be closed, but in principle the server could decide to close it for whatever reason, precisely like the one we're introducing here.

Reply:
I don't disagree with what it does. What I say is RangeFeedRetryError_REASON_RANGEFEED_CLOSED is a retryable error type which is triggered by explicit request to stop rangefeed by client. For me retryable error means someone needs to retry the action.

Reply:
The client needs to retry this if it itself didn't initiate the close. For example in this case here. If the client did decide to close it, then it doesn't care about the error.